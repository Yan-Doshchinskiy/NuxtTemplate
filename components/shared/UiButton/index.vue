<template>
  <button
    :data-selector="`ACTION-${dataSelector}`"
    :disabled="disabled"
    :type="type"
    class="ui-button"
    :class="buttonClass"
    @click="onClick"
    @submit="onSubmit"
    @reset="onReset"
  >
    <UiLoaderSmall v-if="isLoading" :variant="variant" />
    <div v-else class="ui-button__wrapper">
      <div
        v-if="$scopedSlots.iconLeft"
        :class="getIconClass('left')"
      >
        <slot name="iconLeft" />
      </div>

      <span v-if="label" class="ui-button__content">
        {{ label }}
      </span>
      <div
        v-if="$scopedSlots.iconRight"
        :class="getIconClass('right')"
      >
        <slot name="iconRight" />
      </div>
    </div>
  </button>
</template>

<script lang="ts">
import { Component, Emit, Prop, Vue } from 'vue-property-decorator';

import UiLoaderSmall from '~/components/shared/UiLoaderSmall/index.vue';

import type { COMPUTED_STYLE } from '~/enums/ui';

export type TButtonVariant = 'primary' | 'secondary' | 'danger';
export type TIconPosition = 'left' | 'right'

@Component({
  name: 'UiButton',
  components: {
    UiLoaderSmall
  }
})
export default class UiButton extends Vue {
  @Prop({ required: true }) readonly dataSelector!: string;
  @Prop({ default: false }) readonly isLoading!: boolean;
  @Prop({ default: false }) readonly disabled!: boolean;
  @Prop({ default: '' }) readonly label!: string;
  @Prop({ default: 'button' }) readonly type!: string;
  @Prop({ default: 'primary' }) readonly variant!: TButtonVariant;

  private get buttonClass(): COMPUTED_STYLE {
    const { variant } = this;
    return [
      {
        'ui-button_primary': variant === 'primary',
        'ui-button_secondary': variant === 'secondary',
        'ui-button_danger': variant === 'danger',
        'ui-button_loading': this.isLoading
      }
    ];
  }

  private getIconClass(position: TIconPosition): COMPUTED_STYLE {
    return [
      'ui-button__icon',
      {
        'ui-button__icon_big': !this.label,
        'ui-button__icon_left': position === 'left',
        'ui-button__icon_right': position === 'right'
      }
    ];
  }

  @Emit('click')
  private onClick(): void { }

  @Emit('submit')
  private onSubmit(): void { }

  @Emit('reset')
  private onReset(): void {}
}
</script>

<style lang="scss" scoped>
.ui-button {
  width: 100%;
  border-radius: 4px;
  padding: 12px 16px;
  transition: $transition;
  font-weight: 700;
  font-size: 16px;
  line-height: 20px;
  overflow: hidden;
  cursor: pointer;
  border: none;
  height: max-content;
  &__wrapper {
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    overflow: hidden;
    column-gap: 10px;
  }

  &__icon {
    display: flex;
    align-items: center;
    transition: $transition;

    ::v-deep i {
      font-size: 16px;
    }

    &_big {
      ::v-deep i {
        font-size: 20px;
      }
    }

    ::v-deep i::before {
      transition: $transition;
    }
  }

  &__content {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &_primary {
    background-color: $blue;
    color: $white;

    ::v-deep i::before {
      transition: $transition;
      color: $white;
    }

    &:hover {
      background-color: $blue-hover;
    }

    &:active {
      background-color: $blue-hover;
      box-shadow: inset 4px 4px 4px rgba(0, 0, 0, 0.2);
    }

    &:disabled {
      background-color: $gray-500;
      cursor: default;
      box-shadow: none;
    }
  }

  &_secondary {
    color: $blue;
    background-color: transparent;
    border: 1px solid $blue;
    font-weight: 400;
    padding: 11px 15px;
    ::v-deep i::before {
      transition: $transition;
      color: $blue;
    }

    &:hover {
      border-color: $blue-hover;
      color: $blue-hover;

      ::v-deep i::before {
        color: $blue-hover;
      }
    }

    &:active {
      border-color: $blue-hover;
      color: $black;
      box-shadow: inset 4px 4px 4px rgba(0, 0, 0, 0.1);

      ::v-deep i::before {
        color: $blue-hover;
      }
    }

    &:disabled {
      border-color: $gray-500;
      color: $gray-500;
      cursor: default;
      box-shadow: none;

      ::v-deep i::before {
        color: $gray-500;
      }
    }
  }

  &_danger {
    background-color: $red;
    color: $white;

    ::v-deep i::before {
      transition: $transition;
      color: $white;
    }

    &:hover {
      background-color: $red-hover;
    }

    &:active {
      background-color: $red-hover;
      box-shadow: inset 4px 4px 4px rgba(0, 0, 0, 0.2);
    }

    &:disabled {
      background-color: $gray-500;
      cursor: default;
      box-shadow: none;
    }
  }

  &_loading {
    pointer-events: none;
  }

  &_text {
    padding: 0;
    border: none;
    color: $blue;
  }
}
</style>
