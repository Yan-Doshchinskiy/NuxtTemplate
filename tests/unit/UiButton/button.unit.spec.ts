import UiButton from '~/components/shared/UiButton/index.vue';

import MainUnit from '~/tests/unit/prototypes/main.unit';

import MainOptionsUtil from '~/tests/unit/prototypes/utils/main.options.util';
import ButtonStylePart from '~/tests/unit/UiButton/parts/button.style.part';
import ButtonStylerUtil from '~/tests/unit/UiButton/utils/button.styler.util';
import ButtonEmitsPart from '~/tests/unit/UiButton/parts/button.emits.part';
import ButtonAttributesPart from '~/tests/unit/UiButton/parts/button.attributes.part';
import ButtonConditionsPart from '~/tests/unit/UiButton/parts/button.conditions.part';
import ButtonPropsPart from '~/tests/unit/UiButton/parts/button.props.part';

import type { IWrapperOptions } from '~/tests/unit/prototypes/utils/main.options.util';
import type { TButtonVariant } from '~/components/shared/UiButton/index.vue';
import type { Slots } from '@vue/test-utils';
import ButtonMounterUtil from '~/tests/unit/UiButton/utils/button.mounter.util';

export interface IComponentDefaultProps {
  isLoading: boolean,
  disabled: boolean,
  label: string,
  type: string,
  variant: TButtonVariant,
}

export interface IComponentProps extends Partial<IComponentDefaultProps> {
  dataSelector: string,
}

export interface IComponentSlots extends Slots{
  iconLeft: string,
  iconRight: string
}
export type TButtonWrapperOptions = IWrapperOptions<IComponentProps, IComponentSlots>
export type TOptionUtil = MainOptionsUtil<IComponentDefaultProps, IComponentProps, IComponentSlots>;
export type TStylerUtil = ButtonStylerUtil;
export type TMounterUtil = ButtonMounterUtil;

const defaultComponentProps: IComponentDefaultProps = {
  isLoading: false,
  disabled: false,
  label: '',
  type: 'button',
  variant: 'primary'
};

const defaultTestOptions: TButtonWrapperOptions = {
  propsData: {
    dataSelector: 'TEST'
  }
};

class ButtonUnitSpec extends MainUnit<IComponentDefaultProps, IComponentProps, IComponentSlots, TOptionUtil, TStylerUtil, TMounterUtil> {
  constructor() {
    const optionUtil = new MainOptionsUtil({
      defaultComponentProps,
      defaultTestOptions
    });
    const stylerUtil = new ButtonStylerUtil({ wrapperClass: 'ui-button' });
    const mounterUtil = new ButtonMounterUtil({
      component: UiButton,
      wrapperClass: stylerUtil.wrapperClass,
      defaultTestOptions: optionUtil.defaultTestOptions
    });

    super({
      optionUtil,
      stylerUtil,
      mounterUtil
    });
  }
}

const testInstance = new ButtonUnitSpec();

testInstance.addNewTestPart((utils) => new ButtonPropsPart(utils));
testInstance.addNewTestPart((utils) => new ButtonStylePart(utils));
testInstance.addNewTestPart((utils) => new ButtonEmitsPart(utils));
testInstance.addNewTestPart((utils) => new ButtonAttributesPart(utils));
testInstance.addNewTestPart((utils) => new ButtonConditionsPart(utils));
testInstance.run();

export default ButtonUnitSpec;
