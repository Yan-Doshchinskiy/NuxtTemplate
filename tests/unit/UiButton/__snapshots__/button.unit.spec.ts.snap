// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiButton attributes check dataSelector attribute (custom) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-CUSTOM-SELECTOR"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check dataSelector attribute (default) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check disabled attribute (default) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check disabled attribute (false) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check disabled attribute (true) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  disabled="disabled"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check type attribute (button) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check type attribute (default) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check type attribute (reset) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="reset"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton attributes check type attribute (submit) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="submit"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton display conditions entry wrapper is visible with default props 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton display conditions works correctly when isLoading equals to true 1`] = `
<button
  class="ui-button ui-button_primary ui-button_loading"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-loader"
  >
    <div
      class="ui-loader__ring"
    />
  </div>
</button>
`;

exports[`UiButton display conditions works correctly with label and with both icons 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <div
      class="ui-button__icon ui-button__icon_left"
    >
      <i
        class="icon-arrow-left"
      />
    </div>
     
    <span
      class="ui-button__content"
    >
      
      Custom label
    
    </span>
     
    <div
      class="ui-button__icon ui-button__icon_right"
    >
      <i
        class="icon-arrow-right"
      />
    </div>
  </div>
</button>
`;

exports[`UiButton display conditions works correctly with label and with left icon 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <div
      class="ui-button__icon ui-button__icon_left"
    >
      <i
        class="icon-arrow"
      />
    </div>
     
    <span
      class="ui-button__content"
    >
      
      Custom label
    
    </span>
     
    <!---->
  </div>
</button>
`;

exports[`UiButton display conditions works correctly with label and with right icon 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <span
      class="ui-button__content"
    >
      
      Custom label
    
    </span>
     
    <div
      class="ui-button__icon ui-button__icon_right"
    >
      <i
        class="icon-arrow"
      />
    </div>
  </div>
</button>
`;

exports[`UiButton display conditions works correctly with label and without icons 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <span
      class="ui-button__content"
    >
      
      Custom label
    
    </span>
     
    <!---->
  </div>
</button>
`;

exports[`UiButton display conditions works correctly without label and with both icons 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <div
      class="ui-button__icon ui-button__icon_big ui-button__icon_left"
    >
      <i
        class="icon-arrow-left"
      />
    </div>
     
    <!---->
     
    <div
      class="ui-button__icon ui-button__icon_big ui-button__icon_right"
    >
      <i
        class="icon-arrow-right"
      />
    </div>
  </div>
</button>
`;

exports[`UiButton display conditions works correctly without label and with left icon 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <div
      class="ui-button__icon ui-button__icon_big ui-button__icon_left"
    >
      <i
        class="icon-arrow-left"
      />
    </div>
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton display conditions works correctly without label and with right icon 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <div
      class="ui-button__icon ui-button__icon_big ui-button__icon_right"
    >
      <i
        class="icon-arrow-right"
      />
    </div>
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (with loading) 1`] = `
<button
  class="ui-button ui-button_primary ui-button_loading"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-loader"
  >
    <div
      class="ui-loader__ring"
    />
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (with loading) 2`] = `
<button
  class="ui-button ui-button_secondary ui-button_loading"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-loader"
  >
    <div
      class="ui-loader__ring ui-loader__ring_secondary"
    />
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (with loading) 3`] = `
<button
  class="ui-button ui-button_danger ui-button_loading"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-loader"
  >
    <div
      class="ui-loader__ring"
    />
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (withoutLoading) 1`] = `
<button
  class="ui-button ui-button_primary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (withoutLoading) 2`] = `
<button
  class="ui-button ui-button_secondary"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;

exports[`UiButton style correct wrapper class with different variants (withoutLoading) 3`] = `
<button
  class="ui-button ui-button_danger"
  data-selector="ACTION-TEST"
  type="button"
>
  <div
    class="ui-button__wrapper"
  >
    <!---->
     
    <!---->
     
    <!---->
  </div>
</button>
`;
